### PR TITLE
Python layer contributor tests

### PR DESCRIPTION
**Pull request by menehune23**
Adds tests for python layer contributors

To do:

- [x] `Function`
- [x] `FunctionValidator`
- [x] `InvokerDependencyCache`
- [x] `Invoker`

### CODE REVIEW

File: .github/workflows/codeql-analysis.yml [hunk 1]
```diff
@@ -57,7 +57,7 @@ jobs:
     # Autobuild attempts to build any compiled languages  (C/C++, C#, or Java).
     # If this step fails, then you should remove it and run the build manually (see below)
     - name: Autobuild
-      uses: github/codeql-action/autobuild@v1
+      uses: github/codeql-action/autobuild@v2
```
Comment (menehune23) [inline .github/workflows/codeql-analysis.yml:60]: Fixes CI warning (same for below)

File: buildpacks/Makefile [hunk 2]
```diff
@@ -28,5 +28,5 @@ all .PHONY: buildpacks
 buildpacks.clean:
 clean .PHONY: buildpacks.clean
 
-buildpacks.test:
+buildpacks.tests:
```
Comment (menehune23) [inline buildpacks/Makefile:31]: I think this was a typo

Can someone clarify what this syntax does when it's by itself with nothing after the colon?

File: buildpacks/_shared/shared/command_runner.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
+// Copyright 2021-2022 VMware, Inc.
+// SPDX-License-Identifier: BSD-2-Clause
+
+package shared
```
Comment (menehune23) [inline buildpacks/_shared/shared/command_runner.go:4]: Placing this in `shared` from the start, because it will be a useful pattern for future buildpacks

File: buildpacks/python/python/function.go [hunk 4]
```diff
@@ -4,31 +4,59 @@
...
 
+	invoker     Layer
+	invokerDeps Layer
+
 	funcYamlEnvs map[string]string
 }
 
-func NewFunction(opts ...FunctionOpt) Function {
-	f := Function{}
+func NewFunction(
```
Comment (menehune23) [inline buildpacks/python/python/function.go:45]: Note that initializers named `New...` should return pointers

File: buildpacks/python/python/function.go [hunk 5]
```diff
@@ -43,10 +71,10 @@ func NewFunction(opts ...FunctionOpt) Function {
 		},
 	)
 
-	return f
+	return &f
```
Comment (menehune23) [inline buildpacks/python/python/function.go:74]: Note that initializers named `New...` should return pointers, so I updated this (and all the contributor initializers)

File: buildpacks/python/python/function.go [hunk 6]
```diff
@@ -66,23 +94,48 @@ func (f Function) Contribute(layer libcnb.Layer) (libcnb.Layer, error) {
...
+		}
+
 		return layer, nil
 	})
 }
 
-func (f Function) Name() string {
-	return f.layerContributor.Name
-}
+func (f *Function) validate(moduleName, functionName string) error {
```
Comment (menehune23) [inline buildpacks/python/python/function.go:105]: This logic was moved over from the `FunctionValidationLayer` contributor

File: buildpacks/python/tests/setup.go [hunk 7]
```diff
@@ -50,126 +46,7 @@ func (f HTTPFunction) Generate(path, funcTemplate string) error {
 	return templ.Execute(file, f)
 }
 
-func NewLogger() bard.Logger {
```
Comment (menehune23) [inline buildpacks/python/tests/setup.go:53]: Here and below moved to `shared`

File: buildpacks/python/python/function.go [hunk 8]
```diff
@@ -66,23 +94,48 @@ func (f Function) Contribute(layer libcnb.Layer) (libcnb.Layer, error) {
 			layer.LaunchEnvironment.Default(envName, envValue)
 		}
 
+		if err := f.validate(functionModule, functionName); err != nil {
```
Comment (menehune23) [inline buildpacks/python/python/function.go:97]: Note that we are checking the function module and name that are set on this `Function` instance. This is a slight deviation from the `validation.go` implementation as it has logic to adjust the function module and name based on whether an env var overrides it (see [here](https://github.com/vmware-tanzu/function-buildpacks-for-knative/pull/137/files#diff-af4646e8568af1f9e2296d68edf77f7e7b4f3753c3612c4d0edcd866b05c0cb1L103) and [here](https://github.com/vmware-tanzu/function-buildpacks-for-knative/pull/137/files#diff-af4646e8568af1f9e2296d68edf77f7e7b4f3753c3612c4d0edcd866b05c0cb1L117)). That seems unnecessary, as at the end of the day, what needs to be checked is whatever function module/name is configured in this `Function` layer contributor.

Comment (menehune23) [reply buildpacks/python/python/function.go:97]: Side question: Is my understanding correct that the function class is supposed to be overridable at launch/run time via `MODULE_NAME` and `FUNCTION_NAME`? If that's the case, I'd argue that it still makes sense to check the function class that is currently known to this layer contributor, because it will have no way of telling the future and knowing what a user may set at launch.

File: buildpacks/python/python/function.go (hunk 5 above)
Comment (bbtong) [reply buildpacks/python/python/function.go:74]: What's the rationale for this?

File: .github/workflows/codeql-analysis.yml (hunk 1 above)
Comment (bbtong) [inline .github/workflows/codeql-analysis.yml:60]: Version bumping in Actions should probably be a separate PR.

File: buildpacks/_shared/shared/func_yaml.go [hunk 9]
```diff
@@ -1,7 +1,7 @@
 // Copyright 2021-2022 VMware, Inc.
 // SPDX-License-Identifier: BSD-2-Clause
 
-package java
+package shared
```
Comment (bbtong) [inline buildpacks/_shared/shared/func_yaml.go:4]: If this is not a temporarily generated folder, why are we prefixing it with underscore?

Comment (bbtong) [reply buildpacks/_shared/shared/func_yaml.go:4]: Overall, I'm not a fan of the filestructure now being
```
_shared/
python/
java/
```
and think that if a refactor PR was necessary, we should be refactoring into a proper Go Project. Since we broke standards for a per-language folder structure, is building off that the best idea?

File: buildpacks/_shared/tests/setup.go [hunk 10]
```diff
@@ -15,14 +14,13 @@ import (
 )
 
 func NewLogger() bard.Logger {
-	buf := bytes.NewBuffer(nil)
-	return bard.NewLogger(buf)
+	return bard.NewLogger(io.Discard)
```
Comment (bbtong) [inline buildpacks/_shared/tests/setup.go:17]: @andrew-su I don't remember what we were using `buf` with, but `io.Discard` does have different behavior.

File: buildpacks/java/go.mod [hunk 11]
```diff
@@ -1,11 +1,50 @@
...
-go 1.16
+go 1.19
 
 require (
 	github.com/buildpacks/libcnb v1.26.0
 	github.com/paketo-buildpacks/libpak v1.61.0
-	gopkg.in/yaml.v3 v3.0.1
-	knative.dev/kn-plugin-func v0.19.0
-	knative.dev/pkg v0.0.0-20210902173607-844a6bc45596
+	kn-fn/shared v0.0.0
```
Comment (bbtong) [inline buildpacks/java/go.mod:8]: 🤔

File: buildpacks/java/java/detect.go [hunk 12]
```diff
@@ -9,6 +9,8 @@ import (
 	"github.com/buildpacks/libcnb"
 	"github.com/paketo-buildpacks/libpak"
 	"github.com/paketo-buildpacks/libpak/bard"
+
+	"kn-fn/shared/shared"
```
Comment (bbtong) [inline buildpacks/java/java/detect.go:13]: 🤔 Does this import path seem good as a structure?

File: buildpacks/python/go.mod [hunk 13]
```diff
@@ -47,5 +48,8 @@ require (
 	gopkg.in/inf.v0 v0.9.1 // indirect
 	gopkg.in/warnings.v0 v0.1.2 // indirect
 	gopkg.in/yaml.v2 v2.4.0 // indirect
+	gopkg.in/yaml.v3 v3.0.1 // indirect
 	k8s.io/apimachinery v0.20.7 // indirect
 )
+
+replace kn-fn/shared v0.0.0 => ../_shared
```
Comment (bbtong) [inline buildpacks/python/go.mod:55]: Again, this doesn't feel "right" to me.

**Review by bbtong — COMMENTED** [review]
There's a lot more to review -- I'm not done yet.
However, this PR has too much scope imo. There's 3 PRs here:
1) Version bumping CI
2) Python test addition against a stable main 
3) Refactor of the Buildpacks structure

File: buildpacks/python/python/function.go (hunk 5 above)
Comment (menehune23) [reply buildpacks/python/python/function.go:74]: I believe it's for congruency with the `new` keyword. I can't remember exactly where I read it (maybe somewhere in Effective Go), but I heard that most constructors should return pointers, and non-pointer constructors should start with `Make...` (this might help: https://stackoverflow.com/a/18125682/2456938)

Comment (menehune23) [reply buildpacks/python/python/function.go:74]: In our case, we actually want pointers anyway (to pass into the layers array), so it works nicely

File: .github/workflows/codeql-analysis.yml (hunk 1 above)
Comment (menehune23) [reply .github/workflows/codeql-analysis.yml:60]: That's fair. Good catch -- I meant to at least put it in it's own commit. I'll go ahead and move to a separate PR for posterity

File: buildpacks/_shared/shared/func_yaml.go (hunk 9 above)
Comment (menehune23) [reply buildpacks/_shared/shared/func_yaml.go:4]: The underscore was just to differentiate it from the buildpacks, since it's not a buildpack itself. Happy to take suggestions on other names or locations. I considered moving it outside the `buildpacks` dir, but I also wanted it to be clear that it is shared _buildpack_ code, vs general shared code.

File: buildpacks/_shared/tests/setup.go (hunk 10 above)
Comment (menehune23) [reply buildpacks/_shared/tests/setup.go:17]: So far, the tests haven't needed the actual log output so it doesn't make a difference for our current use. 

That said, if a test ever wanted to assert on the log contents, they would need to call `bard.NewLogger(buf)` themselves, then make assertions on `buf.String()`. So I gather that this is just a convenience function to create a new logger that satisfies constructors.

File: buildpacks/java/go.mod (hunk 11 above)
Comment (menehune23) [reply buildpacks/java/go.mod:8]: I turns out it just needs to be any semver. I don't have a preference as to what value we use, since it's purely internal to this repo.

File: buildpacks/java/java/detect.go (hunk 12 above)
Comment (menehune23) [reply buildpacks/java/java/detect.go:13]: I wondered the same. I was following what I saw in our other go subprojects in this repo, which was to have a structure like:

```
<project dir>/
    <primary package>/
    tests/
```

where the main code lives under `<primary package>/`. I originally had it where the main code was up one level, and the import looked a little nicer (`kn-fn/shared`). It's just aesthetics though, so I prioritized what I saw as our convention.

Happy to go either way though. No preference from me.

File: buildpacks/python/go.mod (hunk 13 above)
Comment (menehune23) [reply buildpacks/python/go.mod:55]: To be specific, are you referring to the semver?

**Comment by menehune23** [comment]
> There's a lot more to review -- I'm not done yet. However, this PR has too much scope imo. There's 3 PRs here:
> 
> 1. Version bumping CI
> 2. Python test addition against a stable main
> 3. Refactor of the Buildpacks structure

@bbtong 

For 1, sure I can move it out.

2 & 3 are slightly tied (though not undoable) as I created the shared project as a result of some necessary refactors so that we could test using dependency injection (namely, the creation of a `CommandRunner` interface). Technically, that interface doesn't _have_ to live in a shared place yet, but it wasn't a big lift to have it shared from the get-go since it's not specific to python. Moving `FuncYaml` flowed naturally from that since it is also not buildpack specific. I do get though that it's not strictly in scope of this PR though. Happy to address as a separate PR if it's important for us to separate. It's mostly organizational and not new code, so it shouldn't be too bad either way.

**Comment by menehune23** [comment]
CodeQL workflow changes move to PR: https://github.com/vmware-tanzu/function-buildpacks-for-knative/pull/138

**Comment by menehune23** [comment]
Note that I may need to separate the validation and function layers after all, as the function layer is a launch layer and is therefore re-used during builds. We need the validation to run every build.

File: .github/workflows/codeql-analysis.yml (hunk 1 above)
Comment (menehune23) [reply .github/workflows/codeql-analysis.yml]: Moved to separate PR: https://github.com/vmware-tanzu/function-buildpacks-for-knative/pull/138

File: buildpacks/python/python/function.go (hunk 6 above)
Comment (menehune23) [reply buildpacks/python/python/function.go]: No longer applies

File: buildpacks/python/python/function.go (hunk 8 above)
Comment (menehune23) [reply buildpacks/python/python/function.go]: No longer part of this file. Will move commentary to `validation.go`

File: .github/workflows/codeql-analysis.yml (hunk 1 above)
Comment (menehune23) [reply .github/workflows/codeql-analysis.yml]: PR: https://github.com/vmware-tanzu/function-buildpacks-for-knative/pull/138

File: buildpacks/java/java/detect.go (hunk 12 above)
Comment (menehune23) [reply buildpacks/java/java/detect.go:13]: @bbtong @andrew-su @joeeltgroth any preferences here? Happy to follow whatever the consensus is.

Comment (menehune23) [reply buildpacks/java/java/detect.go:13]: Another option is to name the module itself something other than `kn-fn/shared`, then keep the current folder structure. For instance, if the module were named `kn-fn/buildpacks`, the import could be `kn-fn/buildpacks/shared`. Again, it's mostly aesthetics though, so not sure how much we care.

**Comment by menehune23** [comment]
> Note that I may need to separate the validation and function layers after all, as the function layer is a launch layer and is therefore re-used during builds. We need the validation to run every build.

Heads up, I did go ahead and split them back up as they originally were.

File: buildpacks/Makefile (hunk 2 above)
Comment (bbtong) [reply buildpacks/Makefile:31]: @andrew-su

File: buildpacks/python/python/function.go (hunk 5 above)
Comment (bbtong) [reply buildpacks/python/python/function.go]: I think we need to consider here if we really need it to be a reference instead of a value for our usage; else, we the function is essentially a `make` in its old state.

File: buildpacks/python/python/validation.go [hunk 14]
```diff
@@ -13,117 +12,92 @@ import (
...
-		if override || (vl.module == "" && vl.function == "") {
-			vl.module = fSplit[0]
-			vl.function = fSplit[1]
-			metadata[EnvModuleName] = fSplit[0]
-			metadata[EnvFunctionName] = fSplit[1]
-		}
-	}
-}
-
-func WithValidationFunctionEnvs(envs map[string]any) FunctionValidationOpts {
```
Comment (menehune23) [inline buildpacks/python/python/validation.go:117]: In going to write the tests for this, I came across what I believe to be a simplification. Correct me if I'm wrong, but this class doesn't seem to need to worry about "override" functionality. The only function that it can know to check is based on the function module/name given to it. It shouldn't need to worry about whether the value was a default or not, like the `Function` layer contributor does.

For instance, if the function module/name given is a default. i.e. "override" is false, this would _still_ need to validate it, because there is no telling what function could be run at launch

File: buildpacks/_shared/shared/func_yaml.go (hunk 9 above)
Comment (bbtong) [reply buildpacks/_shared/shared/func_yaml.go:4]: In Go, that underscore would mean that the folder/file should be ignored. I still personally don't feel like 3 shared functions extracted warrants the generation of an internal dependency, which also feels a bit against Go, but alas I'll let others comment for approving.

File: buildpacks/python/go.mod (hunk 13 above)
Comment (bbtong) [reply buildpacks/python/go.mod:55]: Referring to the need for replace with a relative path.

File: buildpacks/_shared/shared/func_yaml.go (hunk 9 above)
Comment (menehune23) [reply buildpacks/_shared/shared/func_yaml.go:4]: Gotcha. Didn't realize that about the underscore. Have an alternative suggestion?

For the shared functions, I would focus on the usage of the individual functions rather than the total number of them. There may be just a handful of small shared functions, but for each one we are using them for N buildpacks.

File: buildpacks/python/python/function.go (hunk 5 above)
Comment (menehune23) [reply buildpacks/python/python/function.go]: That's a valid point. And I think in this case, we could get away with either. I noticed that some of our contributors had pointer receivers and others didn't, so I'm aiming mostly for consistency. The contributors are stateful, so I leaned toward pointer receivers.

Comment (bbtong) [reply buildpacks/python/python/function.go]: Yeah, we have a lack of consistency on that. I agree that we can lean towards pointer-receivers going forward.

File: buildpacks/python/go.mod (hunk 13 above)
Comment (menehune23) [reply buildpacks/python/go.mod:55]: I understand, since typically `replace` statements are not ever checked in. In my research for how to share projects like this, this was the common pattern (and exception to the not-checked-in rule of thumb).

Comment (menehune23) [reply buildpacks/python/go.mod:55]: I guess it comes down to whether we like this more than having to maintain a separate repo (which I gather is something we really want to avoid)

File: buildpacks/python/python/validation.go (hunk 14 above)
Comment (menehune23) [reply buildpacks/python/python/validation.go:117]: I do notice that the previous validator implementation supported overriding the passed in function module/name via the `MODULE_NAME` and `FUNCTION_NAME` env vars from `func.yaml`. This seems inconsistent with `function.go` which does not allow this. Is my understanding correct in that this is an inconsistency that should be corrected, or is there something I'm missing? If the former, which way should we go? (either make the function layer contributor support overriding, or remove that functionality from the validator [which is what's in this PR currently]).

File: buildpacks/java/java/detect.go (hunk 12 above)
Comment (andrew-su) [reply buildpacks/java/java/detect.go:13]: `kn-fn/shared/shared` is definitely a bit awkward to say. I'm okay with either being just `kn-fn/shared` or `kn-fn/buildpacks/shared` with preference for the latter. It's more explicit that it's for the buildpacks (though we don't currently use Go elsewhere)

File: buildpacks/python/go.mod (hunk 13 above)
Comment (andrew-su) [reply buildpacks/python/go.mod:55]: There's very minimal code right now that it's probably not worth throwing into its own repo.

File: buildpacks/python/python/validation.go (hunk 14 above)
Comment (andrew-su) [reply buildpacks/python/python/validation.go:117]: The layer should be running whether something is defined (or not) for `BP_FUNCTION` or in `func.yaml`.

`function.go`'s module precedence:
1. `BP_FUNCTION` if set (override is set to true)
1. `func.yaml`'s `MODULE_NAME` and `FUNCTION_NAME` (Overrides the default env)
1. `BP_FUNCTION` default (nothing to override it so here we are)

Is that not the overriding behaviour you're seeing?

Perhaps to simplify the layers, we should process what the function name and module should be once at the start of the build. These names can be passed to each layer that needs it instead of having each layer translate it.

File: buildpacks/_shared/shared/func_yaml.go (hunk 9 above)
Comment (bbtong) [reply buildpacks/_shared/shared/func_yaml.go:4]: [This article](https://dave.cheney.net/2019/01/08/avoid-package-names-like-base-util-or-common) summarizes kinda what I'm getting at. Golang's [direct blog](https://go.dev/blog/package-names?utm_source=dlvr.it&utm_medium=twitter) also says to directly avoid this. We basically have a `utils` or `helpers` folder now sitting, so we're either breaking repetition or we're breaking structure.

Definitely let's remove the `_` underscore prefix. If I _had_ to name it, I would follow their guidance and name the package relevant to what it does: it handles `func.yaml`.

File: buildpacks/Makefile (hunk 2 above)
Comment (andrew-su) [reply buildpacks/Makefile:31]: Oops, thought I responded to this, sorry!

This was added here to let you know that this target should be inside every buildpack (additionally helps with autocomplete). Think of it like a placeholder.

File: buildpacks/python/go.mod (hunk 13 above)
Comment (bbtong) [reply buildpacks/python/go.mod:55]: Did not mean to make a separate repo, was just wondering if there's a different way to handle `replace`.

File: buildpacks/python/python/validation.go (hunk 14 above)
Comment (bbtong) [reply buildpacks/python/python/validation.go:117]: I don't recall the history on this fully, so I'll defer this to @andrew-su. 

Overriding via environment variables was intentional for a previous feature ask though. IIRC `function.go` does also override, though?

File: buildpacks/_shared/go.mod [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+go 1.19
+
+require (
+	github.com/paketo-buildpacks/libpak v1.61.0
+	gopkg.in/yaml.v3 v3.0.1
+	knative.dev/kn-plugin-func v0.19.0
+	knative.dev/pkg v0.0.0-20210902173607-844a6bc45596
+)
+
+require (
```
Comment (bbtong) [inline buildpacks/_shared/go.mod:12]: Just doublechecking that a `go mod tidy` was run on all of your `go.mod`'s?

File: buildpacks/python/python/validation.go [hunk 16]
```diff
@@ -13,117 +12,92 @@ import (
...
+		Cache: false,
 	})
 
-	return fvl
+	return fv
 }
 
-func (i FunctionValidationLayer) Contribute(layer libcnb.Layer) (libcnb.Layer, error) {
-	i.layerContributor.Logger = i.logger
+func (fv *FunctionValidator) Contribute(layer libcnb.Layer) (libcnb.Layer, error) {
```
Comment (bbtong) [inline buildpacks/python/python/validation.go:69]: Just wanted to call out this is a good refactor `i` --> `fv`, but you could definitely make a PR of _just_ refactors in the future to avoid a multi-hundred (or even multi-thousand) line PR and get this type of change in sooner.

File: buildpacks/python/python/validation.go (hunk 14 above)
Comment (andrew-su) [reply buildpacks/python/python/validation.go:117]: Essentially, `MODULE_NAME` and `FUNCTION_NAME` is an implementation detail of the invoker and isn't a part of our buildpack's API. They may choose to use it, but can't guarantee it'll still be the same or around in the future.

File: buildpacks/_shared/go.mod (hunk 15 above)
Comment (menehune23) [reply buildpacks/_shared/go.mod:13]: Good suggestion, done!

File: buildpacks/_shared/shared/func_yaml.go (hunk 9 above)
Comment (menehune23) [reply buildpacks/_shared/shared/func_yaml.go:4]: Ah, I see what you're getting at. I'm with you on package names for sure and will address this.

That said, the package name is a separate concern from the folder (and/or module) name that houses these shared packages, so what do you think of this?

```
buildpacks/
    common/
        go.mod  # module name "kn-fn/buildpacks", per later thread in this PR
        func/  # package to house FuncYaml
        command/  # package to house `command.Runner` interface and default implementation
        ...  # other packages as necessary over time
    python/
    java/    
```

cc: @andrew-su @joeeltgroth

File: buildpacks/java/java/detect.go (hunk 12 above)
Comment (menehune23) [reply buildpacks/java/java/detect.go:13]: Will change module name to `kn-fn/buildpacks` per this thread: https://github.com/vmware-tanzu/function-buildpacks-for-knative/pull/137/files#r987176488

File: buildpacks/python/go.mod (hunk 13 above)
Comment (menehune23) [reply buildpacks/python/go.mod:55]: Yeah figured you didn't want that either 🙂 . I meant that from what I've researched, that's the only other option so I think we're stuck with `replace` for this use case. I could be wrong, but I haven't seen any other ways thus far.

File: buildpacks/python/python/validation.go (hunk 14 above)
Comment (menehune23) [reply buildpacks/python/python/validation.go:117]: @bbtong The reason it comes into play in this PR is that I'm writing tests for it and it's currently very complex. I didn't want to invest the time to write tests for it if it is actually behavior we don't want to enforce.

File: buildpacks/_shared/shared/func_yaml.go (hunk 9 above)
Comment (andrew-su) [reply buildpacks/_shared/shared/func_yaml.go:4]: I think either `common` or `shared` is fine (without the `_`). The readme in this folder could be used to hold the context on which the folders are. Would you be okay with pulling this into its own PR? I see your comment further down about DI. Could this be something that is added to the shared folder as a followup refactor?

File: buildpacks/python/python/validation.go (hunk 14 above)
Comment (menehune23) [reply buildpacks/python/python/validation.go:117]: @andrew-su In looking at the code on `main` and also experimenting locally, this is how I understand the current behavior (before this PR):

`function.go` will take in the function class (usually given by `BP_FUNCTION`), separate it at the `.`, and store it in the launch envs as `MODULE_NAME` and `FUNCTION_NAME` (setting it as either a default or an override).

`validation.go` will also accept the same function class (via `BP_FUNCTION`), separate it at the `.`, and store it in its struct. Additionally though, it will allow the user to override the module/name via `MODULE_NAME` and `FUNCTION_NAME` if specified in `func.yaml`.

This additional behavior is what I'm flagging as inconsistent and unnecessary (and makes testing harder/more complex).

A concrete example: if `BP_FUNCTION` is unset and `func.yaml` defines `MODULE_NAME=func` `FUNCTION_NAME=other`, `function.go` will use `func.main` (the default) and `validation.go` will validate `func.other`.

Comment (menehune23) [reply buildpacks/python/python/validation.go:117]: @bbtong I do understand that its arguable that this is a potential bug and should be in a separate PR. From a practical standpoint, it was highly related to the new tests I was adding. But if we'd like to separate this discussion out to a separate PR, I can skip testing `validation.go` for now and do that refactor/testing as a separate PR. Thoughts?

File: buildpacks/_shared/shared/func_yaml.go (hunk 9 above)
Comment (menehune23) [reply buildpacks/_shared/shared/func_yaml.go:4]: Sounds good. I'll spend some time to simplify this PR and will spin out further stories/PRs for the other work

File: buildpacks/java/java/detect.go (hunk 12 above)
Comment (menehune23) [reply buildpacks/java/java/detect.go]: Removed shared code for this PR and will address in later PR

File: buildpacks/python/python/validation.go (hunk 14 above)
Comment (menehune23) [reply buildpacks/python/python/validation.go]: Spoke with @andrew-su to clarify expected behavior. Will address any bugs (if there actually is a bug) in a later PR. We also talked about just removing the need for the `MODULE_NAME` and `FUNCTION_NAME` env vars via invoker refactors, but we can discuss later.

File: buildpacks/python/python/invoker.go [hunk 17]
```diff
@@ -20,12 +20,12 @@ type Invoker struct {
 	pythonPath string
 }
 
-func NewInvoker(dependency libpak.BuildpackDependency, cache libpak.DependencyCache) Invoker {
+func NewInvoker(dependency libpak.BuildpackDependency, cache libpak.DependencyCache) *Invoker {
```
Comment (menehune23) [inline buildpacks/python/python/invoker.go:23]: It would have been nice to supply a mock for the `libpak.DependencyCache` so that we don't need actual testdata on the filesystem (see `testdata/` dir), but it's not possible with the current state of the code, mostly due to the way `libpak.DependencyLayerContributor` is designed -- it takes a concrete type, not an interface.

Still possible to address by a layer of abstraction, but not worth the refactor for this PR

**Comment by menehune23** [comment]
@andrew-su @bbtong I've reworked this PR quite a bit to be more test-focused, as it was originally intended. It should be quite a bit more manageable now 🚀

**Review by bbtong — APPROVED** [review]
LGTM 👍 Output in the GitHub Action's `Test Buildpacks` log looks nice and readable too.

File: buildpacks/python/python/function.go [hunk 18]
```diff
@@ -27,12 +27,12 @@ type Function struct {
 	funcYamlEnvs map[string]string
 }
 
-func NewFunction(opts ...FunctionOpt) Function {
-	f := Function{}
+func NewFunction(opts ...FunctionOpt) *Function {
```
Comment (andrew-su) [inline buildpacks/python/python/function.go:30]: Why'd we switch to using pointers for the layers? I feel like you might have answered it in another comment but I couldn't find it.

Comment (menehune23) [reply buildpacks/python/python/function.go:30]: Here's the original thread. In summary, there are a few reasons:

- Convention: The term "new" sort of implies pointers (vs "make") -- details/links in that thread
- Consistency: Some of our layer contributors were already doing this, and one of them even had pointer receivers but the constructor was not returning a pointer

I don't feel too strongly which side we land on in the consistency bullet, as long as it's consistent. I tend to lean toward pointer receivers in this case because the contributors in general are somewhat stateful.

Comment (menehune23) [reply buildpacks/python/python/function.go:30]: But the functional programmer in me hates stateful objects, so if we want to veer away from them and embrace FP as much as possible, I'm for it too! 😁

Comment (andrew-su) [reply buildpacks/python/python/function.go:30]: As long as we're consistent in the usages, I'm fine with either. Thanks!

**Review by andrew-su — APPROVED** [review]
These changes look great. Thanks for doing this!
